File: day01/solution.go
package main

import ("bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"cmiles74/util")

func load_sample() ([]int, []int){
	return []int{3, 4, 2, 1, 3, 3}, []int{4, 3, 5, 3, 9, 3}
}

func load_input(filename string) ([]int, []int) {
	file, error := os.Open(filename)
	if error != nil {
		panic(error)
	}
	defer file.Close()

	var list_1 = []int{}
	var list_2 = []int{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		values := strings.Fields(scanner.Text())

		value_1, error_1 := strconv.Atoi(values[0])
		value_2, error_2 := strconv.Atoi(values[1])
		if error_1 != nil || error_2 != nil {
			if error_1 != nil {
				panic(error_1)
			} else if error_2 != nil {
				panic(error_2)
			}
		}

		list_1 = append(list_1, value_1)
		list_2 = append(list_2, value_2)
	}

	if error := scanner.Err(); error != nil {
		panic(error)
	}

	return list_1, list_2
}

func sort_list(list []int) {
	sort.Slice(list, func(i, j int) bool {
		return list[i] < list[j]
	})
}

func count_occurs(sorted_list []int, search_value int) (int) {
	occurs := 0
	for _, value := range sorted_list {
		if value == search_value {
			occurs += 1
		}

		if value > search_value {
			break;
		}
	}
	return occurs
}

func part_1() {
	//list_1, list_2 := load_sample()
	list_1, list_2 := load_input("input.txt")
	sort_list(list_1)
	sort_list(list_2)

	sum_distance := 0
	for index, value := range list_1 {
		value_1 := value
		value_2 := list_2[index]
		sum_distance += util.AbsDiff(value_1, value_2)
	}

	fmt.Println(sum_distance)
}

func part_2() {
	//list_1, list_2 := load_sample()
	list_1, list_2 := load_input("input.txt")
	sort_list(list_2)

	sum_occurs := 0
	for _, value := range list_1 {
		sum_occurs += (value * count_occurs(list_2, value))
	}

	fmt.Println(sum_occurs)
}

func main() {
	fmt.Println("Part 1 - Sum of Distances:")
	defer util.Timer("Part 1")()
	part_1()

	fmt.Println("\nPart 2 - Sum of Occurences:")
	defer util.Timer("Part 2")()
	part_2()
	fmt.Println("\n----")
}

File: day01/java/solution.java
package cmiles74.adventofcode;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.InvalidParameterException;
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.concurrent.Callable;

class Solution {

    int[][] loadSample() {
        return new int[][] {
            {3, 4, 2, 1, 3, 3},
            {4, 3, 5, 3, 9, 3}
        };
    }

    int[][] loadInput(String filename) throws IOException {
        var list1 = new ArrayList<Integer>();
        var list2 = new ArrayList<Integer>();

        var lines = Files.readAllLines(Paths.get(filename));
        for (String line : lines) {
            String[] values = line.split("\\s+");
            list1.add(Integer.parseInt(values[0].trim()));
            list2.add(Integer.parseInt(values[1].trim()));
        }

        return new int[][] {
            list1.stream().mapToInt(i -> i).toArray(),
            list2.stream().mapToInt(i -> i).toArray()
        };
    }

    public static void main (String[] args) throws Exception {
        Solution solution = new Solution();
        //int[][] input = solution.loadSample();
        int[][] input = solution.loadInput("../input.txt");

        System.out.println("Part 1 - Sum of Distances:");
        var timer = new Timer<Integer>(new Part1(input));
        var result = timer.time();
        System.out.println(result.getResult());

        System.out.println("Part 2 - Sum of Occurrences:");
        var timer2 = new Timer<Integer>(new Part2(input));
        var result2 = timer2.time();
        System.out.println(result2.getResult());

        System.out.println("\n----");
        System.out.println("Part 1 completed " + result.getElapsed() + " ms");
        System.out.println("Part 2 completed " + result2.getElapsed() + " ms");
    }
}

class Part1 implements Callable<Integer> {
    int[] list1;
    int[] list2;

    Part1(int[][] input) {
        list1 = input[0];
        list2 = input[1];
    }

    public Integer call() throws InvalidParameterException {
        Arrays.sort(list1);
        Arrays.sort(list2);

        int distance = 0;
        for (int index = 0; index < list1.length; index++) {
            distance += Math.abs(list1[index] - list2[index]);
        }

        return distance;
    }
}

class Part2 implements Callable<Integer> {
    int[] list1;
    int[] list2;

    Part2(int[][] input) {
        list1 = input[0];
        list2 = input[1];
    }

    int countOccurs(int[] list, int searchValue) {
        int occurs = 0;
        for (int value : list) {
            if (value == searchValue) {
                occurs += 1;
            }

            if (value > searchValue) {
                break;
            }
        }

        return occurs;
    }

    public Integer call() throws InvalidParameterException {
        Arrays.sort(list2);

        int sumOccurs = 0;
        for (int value : list1) {
            sumOccurs += (value * countOccurs(list2, value));
        }

        return sumOccurs;
    }
}

class TimedResult<Double, T> {
    Double elapsed;
    T result;

    TimedResult(Double elapsed, T result) {
        this.elapsed = elapsed;
        this.result = result;
    }

    public Double getElapsed() {
        return elapsed;
    }

    public T getResult() {
        return result;
    }
}

class Timer<T> {
    Callable<T> callable;

    Timer(Callable<T> callable) {
        this.callable = callable;
    }

    public TimedResult<Double, T> time() throws Exception {
        var start = System.nanoTime();
        T result =  callable.call();
        var stop = System.nanoTime();
        var elapsed = (stop - start) / 1000000.0;
        return new TimedResult<Double, T>(elapsed, result);
    }
}
